daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala

```diff
@@ -454,30 +454,40 @@ class CLIConf(arguments: Array[String]) extends scallop.ScallopConf(arguments) {
     footer("""|
               |Run 'daffodil generate <language> --help' for subcommand specific options""".stripMargin)
 
-    object c extends scallop.Subcommand("c") {
-      banner("""|Usage: daffodil generate c -s <schema> [-r <root>]
-                |                           [-c <file>] [-T<tunable>=<value>...]
-                |                           [outdir]
-                |
-                |Generate C code from a DFDL schema to parse or unparse data
-                |
-                |Generate Options:""".stripMargin)
-
-      descr("Generate C code from a DFDL schema")
+    // Takes language by name so we can pass it to scallop.Subcommand and interpolate it
+    // into strings without getting a puzzling java.lang.ClassCastException at runtime (class
+    // scala.collection.mutable.WrappedArray$ofRef cannot be cast to class java.lang.String)
+    class LanguageConf(languageArg: => String) extends scallop.Subcommand(languageArg) {
```

Comment:
I just tested this with

```scala
    object rust extends LanguageConf("rust") {
      override val languageName = "Rust"
    }
```

But `daffodil generate rust --help` outputs this:

> Generate null code from a DFDL schema to parse or unparse data

I wonder if this needs to use an early initializer? Or maybe it needs to be def or lazy val?

Reply:
Great catch, Steve.  I looked at pre-initialized values but no thanks, that syntax will puzzle new Scala programmers and they would need examples in order to use that syntax.  I tried both def and lazy val and I found out that lazy val works better in practice.  You need to use def in both places for def to work.  If you use val in the object rust, the Scala compiler won't warn you, but the help message will print "null" again.  However, the Scala compiler will refuse to compile Main if you don't use lazy val in both places; it will tell you `value languageName must be declared lazy to override a concrete lazy value` if you use val and `method languageName needs to be a stable, immutable value` if you use def.  Lazy val it is.

Reply:
Unless scallop.Subcommand class also takes a by-name argument there's no point in taking a by-name argument in your LanguageConf class. 

This may be part of the issue. At the time the superclass is constructed it may not have a value for languageArg yet.

Reply:
> Unless scallop.Subcommand class also takes a by-name argument there's no point in taking a by-name argument in your LanguageConf class. This may be part of the issue. At the time the superclass is constructed it may not have a value for languageArg yet.

Er, the whole point of taking the by-name argument is to avoid a runtime exception which invariably occurs if you pass a lang: String argument and use it in both scallop.Subcommand(lang) and val language = lang.  Note scallop.Subcommand doesn't take a by-name argument, but it takes a String* argument which causes lang to somehow become wrapped in a scala.collection.mutable.WrappedArray$ofRef object at runtime in BOTH places.  Formerly I was passing a literal "c" in both scallop.Subcommand("c") and val language = "c" before, so that's why this exception didn't happen before.  The exception also goes away if LanguageConf takes two redundant lang1: String, lang2: String arguments and passes either lang1 or lang2 to scallop.Subcommand(lang1) and the other one of lang1 or lang2 to val language = lang2.  If I add an Assert.invariant(lang1 == lang2) line to double check that both lang1 and lang2 were passed the same value ("c"), however, that line will trigger the exception right there.  Passing the string argument by-name still allows the string to be passed to scallop.Subcommand(languageArg) but prevents the string from being wrapped when it's used in val language = languageArg.  This code is as good as we can get until we upgrade to a new Scala dialect which may stop wrapping the lang: String argument at the wrong place or the wrong time.